### ISSUE TITLE
Pulumi preview quits with "an unhandled error occurred: Program exited with non-zero exit code: 1"

### ISSUE THREAD

**Issue by mortaelth** [pulumi/pulumi#9435]
### What happened?

Hello, 

I am receiving error without any furter details when doing preview of deployment of the helm chart ["tempo-distributed" from grafana](https://github.com/grafana/helm-charts/tree/main/charts/tempo-distributed) (tried many versions of the chart, including very old ones, but the currently latest which does not work is .16.9 )

### Steps to reproduce

I have minimized inputs, it is reproducible even without any input values:
```typescript
this.chart = new k8s.helm.v3.Chart(
  "tempo",
  {
    path: path.resolve(__dirname, "../../helm-charts/tempo-distributed"),
    namespace,
  },
  {  parent: this }
 );
```

### Expected Behavior

preview of components to get installed

### Actual Behavior

error: an unhandled error occurred: Program exited with non-zero exit code: 1

### Versions used

nodejs v17.9.0  //tested on v16 as well
@pulumi/pulumi  3.29.1
typescript      4.3.5

### Additional context

I have tried to located the exact place where it is failing to provide additional details, but without success

it is AKS Kubernetes cluster

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by Frassle** [comment]
Can you run with `--logflow --logtostderr -v9` and attach the output from that to this ticket?

**Comment by mortaelth** [comment]
sure, attaching output, thank you
[log.zip](https://github.com/pulumi/pulumi/files/8529468/log.zip)

**Comment by mortaelth** [comment]
Hello @Frassle, is there anything interesting in the logs? We were not able to resolve it, so considering if we should rather use different means to deploy it

**Comment by mortaelth** [comment]
Hello @Frassle, I believe last time process must have exited prematurely due to some other issue. Sorry for that. This time the log is even bigger, but should be complete.

Sorry for double compression, the 7z is much smaller than usual zip (11M vs 49M)
[log_2.zip](https://github.com/pulumi/pulumi/files/8601983/log_2.zip)

I am really wondering what might be the reason of pulumi exiting that way. Thank you again for helping me out.

**Comment by Frassle** [comment]
You've got promise leaks:
```
The Pulumi runtime detected that 242 promises were still active
at the time that the process exited. There are a few ways that this can occur:
  * Not using `await` or `.then` on a Promise returned from a Pulumi API
  * Introducing a cyclic dependency between two Pulumi Resources
  * A bug in the Pulumi Runtime
Leaving promises active is probably not what you want. If you are unsure about
why you are seeing this message, re-run your program with the `PULUMI_DEBUG_PROMISE_LEAKS`
environment variable. The Pulumi runtime will then print out additional
debug information about the leaked promises.
```

Try running with PULUMI_DEBUG_PROMISE_LEAKS set to help with debugging why. I would of thought this would have printed to the console as part of `pulumi up` as well to make it easy to notice rather than trawling logs.

**Comment by mortaelth** [comment]
Normally it is not printed there, only when trying to install the chart. I
thought it might be related to unexpected exit, that promises were not able
to resolve.

I tried that debug option already, will submit log tomorrow.

Jiri

Dne po 2. kvě 2022 13:25 uživatel Fraser Waters ***@***.***>
napsal:

**Comment by mortaelth** [comment]
Attaching log with PULUMI_DEBUG_PROMISE_LEAKS enabled
[logPromises.zip](https://github.com/pulumi/pulumi/files/8608965/logPromises.zip)

**Comment by Frassle** [comment]
I think your code must be doing some odd with promises/outputs. I've tried running that helm-chart by itself and it seems to render out fine without error. Doesn't immediately seem like a Pulumi issue, can you try reducing your code to the smallest program that reproduces this that you can share with us?

**Comment by mortaelth** [comment]
interesting... Calling this simple example will reproduce that on our side

```
import * as pulumi from "@pulumi/pulumi";
import * as k8s from "@pulumi/kubernetes";
import * as path from "path";
import * as _ from "lodash";

import { TempoProps } from "./data";

export class GrafanaTempo extends pulumi.ComponentResource {
    chart: k8s.helm.v3.Chart;
    constructor(name: string, args: TempoProps, opts?: pulumi.ComponentResourceOptions) {
        super("tracing", name, args, opts);

        this.chart = new k8s.helm.v3.Chart(
            "tempo",
            {
                path: path.resolve(__dirname, "../../helm-charts/tempo-distributed"),
                namespace: "tempo",
            },
            {
                parent: this,
                provider: opts?.provider,
            }
        );
    }
}
```

**Comment by mortaelth** [comment]
Hello @Frassle, it took some time, but here it is. I am attaching sample on which I can reproduce it anytime I run pulumi up. 
Please let me know if you find anything suspicious. If there is anything you would like me to try, I will be ready.
[r110-cat.zip](https://github.com/pulumi/pulumi/files/8660961/r110-cat.zip)

**Comment by Frassle** [comment]
That code doesn't build:
```
grafanaTempo/data.ts:1:39 - error TS2307: Cannot find module '../configInterfaces/ContainerConfigDefinitions' or its corresponding type declarations.

1 import { ExternalContainerConf } from "../configInterfaces/ContainerConfigDefinitions";
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

memoryRatio.ts:1:27 - error TS2307: Cannot find module './configInterfaces/ContainerConfigDefinitions' or its corresponding type declarations.

1 import { Resources } from "./configInterfaces/ContainerConfigDefinitions";
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mortaelth** [comment]
Sorry for that, submitting corrected zip. Strange that pulumi up didnt protest.

[r110-cat-tempo-error.zip](https://github.com/pulumi/pulumi/files/8667167/r110-cat-tempo-error.zip)

**Comment by mortaelth** [comment]
Hello @Frassle, have you been able to detect anything from the code? Hope it was working correctly

**Comment by Frassle** [comment]
Sorry I've been away for a couple of weeks. I'll try that code out this week.

**Comment by Frassle** [comment]
Sorry finally got around to taking a look at this but it still won't run:
```
    grafanaTempo/data.ts(1,39): error TS2307: Cannot find module '../configInterfaces/ContainerConfigDefinitions' or its corresponding type declarations.
```

Might be easier for you to just send us debug logs. Can you run pulumi with `-v11 --logstderr --logflow` and attach the result of that to this ticket or email to me at fraser at pulumi.com

**Comment by mortaelth** [comment]
Reproduced again,
I have used the index.ts from the attachment, had commented out the helm chart to create the cluster.
And then I have run the 'pulumi up -v11 --logtostderr --logflow 2> log.txt' from which I am attaching the log. It died in the preview part of the pulumi up command.
[log.zip](https://github.com/pulumi/pulumi/files/8888396/log.zip)

**Comment by Frassle** [comment]
So that's an index.ts that looks like:
```
const cluster = new AksCluster("aks");
const ctx = createContext();
const tracingSpace = new KubeNamespaceResource("tracing-namespace", {
    kubeconfig: cluster.cluster.kubeConfigRaw,
    namespace: "tracing",
    networkPolicy: {
        denyIngress: false,
    },
    ctx,
});
```

Does it work if you just make the AksCluster?

**Comment by mortaelth** [comment]
Yes, it does, pulumi up finishes correctly. When I put the helm chart in, it fails during preview part of the "up" command

**Comment by Frassle** [comment]
Can you try removing the `dependsOn: tracingSpace,` from the GrafanTempo options?

**Comment by mortaelth** [comment]
I was able to make it running in the simplified sample. It was missing the provider reference and was causing problems when parent was used in the resources inside the tempo Component.

I will dig in it on my side now, trying to making work in the real cluster.

Thank you for your support so far

**Comment by mortaelth** [comment]
Interesting - the only change needed in order to make it working was to remove parent from the k8s.helm.v3.Chart() resource.

I was specifically looking for for some circular dependency in the parent tree structure, but it looks good. There is no resource which is directly or indirectly parent to its children.

Not sure if it helps anyhow, just sharing what I discovered.

**Comment by Frassle** [comment]
> Not sure if it helps anyhow, just sharing what I discovered.

Yup I think this is a bug in the awaiting logic where depending on and parenting off a resource gets the SDK stuck.

**Comment by mortaelth** [comment]
Yes, it gets created correctly, but when I put the helm chart in, it fails

Dne po 13. čvn 2022 13:41 uživatel Fraser Waters ***@***.***>
napsal:

**Comment by furier** [comment]
I get this error 50% of the time when running my azure pipelines with the following yaml.
This forces me to keep redeploying the failed build until for some reason pulumi decides for them to succeed.

```yml
parameters:
  - name: PULUMI_ACCESS_TOKEN # This is the Pulumi access token
    type: string
  - name: environment # This is the environment name, e.g. dev, prod
    type: string
  - name: stack # This is the name of the stack in Pulumi
    type: string

steps:
  - download: current
    displayName: Download Pulumi Artifact
    artifact: pulumi
  - task: ExtractFiles@1
    displayName: Extract Pulumi Artifact
    inputs:
      archiveFilePatterns: $(Pipeline.Workspace)/pulumi/pulumi.zip
      destinationFolder: $(Pipeline.Workspace)/pulumi-output
      cleanDestinationFolder: true
  - task: Pulumi@1
    displayName: Deploy Infrastructure with Pulumi
    env:
      PULUMI_ACCESS_TOKEN: ${{ parameters.PULUMI_ACCESS_TOKEN }}
    inputs:
      azureSubscription: MySubscription
      command: up
      args: --yes --diff --refresh --non-interactive --skip-preview --logflow
      cwd: $(Pipeline.Workspace)/pulumi-output/infrastructure
      stack: ${{ parameters.stack }}
      createStack: true
```

**Comment by Arttii** [comment]
I am having the same issues as well, sometimes it works and sometimes it doesn't.

**Comment by insider89** [comment]
I have the same as well for parent when migrated to pulumi/eks v3 from pulumi/eks v2
